tools/generators/ethereum/contract_events_template_content.go

```diff
@@ -5,71 +5,137 @@ var contractEventsTemplateContent = `{{- $contract := . -}}
-			End:   endBlock,
-		},
-		{{$event.IndexedFilters}}
-	)
-	if err != nil {
-		return nil, fmt.Errorf(
-			"error retrieving past {{$event.CapsName}} events: [%v]",
-			err,
-		)
+) *{{$event.SubscriptionCapsName}} {
+	if opts == nil {
+		opts = new(ethutil.SubscribeOpts)
+	}
+	if opts.TickDuration == 0 {
+		opts.TickDuration = ethutil.DefaultSubscribeOptsTickDuration
+	}
+	if opts.BlocksBack == 0 {
+		opts.BlocksBack = ethutil.DefaultSubscribeOptsBlocksBack
 	}
 
-	events := make([]*abi.{{$contract.AbiClass}}{{$event.CapsName}}, 0)
-
-	for iterator.Next() {
-		event := iterator.Event
-		events = append(events, event)
+	return &{{$event.SubscriptionCapsName}}{
+		{{$contract.ShortVar}},
+		opts,
+		{{$event.IndexedFilters}}
 	}
+}
 
-	return events, nil
+type {{$event.SubscriptionCapsName}} struct {
+	contract *{{$contract.Class}}
+	opts *ethutil.SubscribeOpts
+	{{$event.IndexedFilterFields -}}
 }
 
-func ({{$contract.ShortVar}} *{{$contract.Class}}) Watch{{$event.CapsName}}(
-	success {{$contract.FullVar}}{{$event.CapsName}}Func,
-	{{$event.IndexedFilterDeclarations -}}
-) (subscription.EventSubscription) {
-	eventOccurred := make(chan *abi.{{$contract.AbiClass}}{{$event.CapsName}})
+type {{$contract.FullVar}}{{$event.CapsName}}Func func(
+	{{$event.ParamDeclarations -}}
+)
 
+func ({{$event.SubscriptionShortVar}} *{{$event.SubscriptionCapsName}}) OnEvent(
+	handler {{$contract.FullVar}}{{$event.CapsName}}Func,
+) subscription.EventSubscription {
+	onEventChan := make(chan *abi.{{$contract.AbiClass}}{{$event.CapsName}})
 	ctx, cancel := context.WithCancel(context.Background())
 
-	// TODO: Watch* function will soon accept channel as a parameter instead
-	// of the callback. This loop will be eliminated then.
 	go func() {
 		for {
 			select {
 			case <-ctx.Done():
 				return
-			case event := <-eventOccurred:
-				success(
-                    {{$event.ParamExtractors}}
+			case event := <- onEventChan:
+			    handler(
+					{{$event.ParamExtractors}}
+				)
+			}
+		}
+	}()
+
+	sub := {{$event.SubscriptionShortVar}}.Pipe(onEventChan)
+	return subscription.NewEventSubscription(func() {
+		sub.Unsubscribe()
+		cancel()
+	})
+}
+
+func ({{$event.SubscriptionShortVar}} *{{$event.SubscriptionCapsName}}) Pipe(
+	sink chan *abi.{{$contract.AbiClass}}{{$event.CapsName}},
+) subscription.EventSubscription {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		ticker := time.NewTicker({{$event.SubscriptionShortVar}}.opts.TickDuration)
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				lastBlock, err := {{$event.SubscriptionShortVar}}.contract.blockCounter.CurrentBlock()
+				if err != nil {
+					{{$logger}}.Errorf(
+						"subscription failed to pull events: [%v]",
+						err,
+					)
+				}
+				fromBlock := lastBlock-{{$event.SubscriptionShortVar}}.opts.BlocksBack
+
+				{{$logger}}.Debugf(
+					"Subscription monitoring fetching past {{$event.CapsName}} events " +
```

Comment:
Same about the lowercase.

Reply:
See a05af3111090eecc167afc6106a863543da812e2 please.